walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,118 @@
+        TimeRange curMandatoryRange = mandatoryTimeRanges.get(mandatoryIndex);
+        TimeRange curOptionalRange = optionalTimeRanges.get(optionalIndex);
+        // if they overlap, need to see if any common timerange can be used
+        TimeRange overlap = getOverlap(curMandatoryRange, curOptionalRange);
+        if (overlap != null && overlap.duration() >= request.getDuration()) {
+            overlapTimeRanges.add(overlap);
+        } 
+        if (curOptionalRange.end() <= curMandatoryRange.end()) {
+            optionalIndex++;
+        }
+        if (curMandatoryRange.end() <= curOptionalRange.end()) {
+            mandatoryIndex++;
+        }
+    }
+    if (MandatoryAttendants.size() != 0 && overlapTimeRanges.size() == 0) {
+        return mandatoryTimeRanges;
+    }    
+    return overlapTimeRanges;
   }
+
+    private TimeRange getOverlap(TimeRange mandatoryRange, TimeRange optionalRange) {
+        if (!mandatoryRange.overlaps(optionalRange)) { 
+            return null;
+        }
+        // if the optional range contains the mandatory range, then the entirety of the mandatory range is shared by all
+        if (optionalRange.contains(mandatoryRange)) {
+            return mandatoryRange;
+        }
+        // if the mandatory range contains the optional range, the entirety of optional range is shared by all
+        if (mandatoryRange.contains(optionalRange)) {
+            return optionalRange;
+        }
+        if (mandatoryRange.start() < optionalRange.start()) {
+            return TimeRange.fromStartEnd(optionalRange.start(), mandatoryRange.end(), true);
+        } else {
+            return TimeRange.fromStartEnd(mandatoryRange.start(), optionalRange.end(), true);
+        }
+    }
+
+    private void handleEventConflict(ArrayList<TimeRange> timeranges, Event e, MeetingRequest request) {
+        for (int i = 0; i < timeranges.size();) {
+            TimeRange curRange = timeranges.get(i);
+            TimeRange eventRange = e.getWhen();
+            if (curRange.overlaps(eventRange)) {
+                // case 1: current timerange is within event timerange => no part of timerange works so remove
+                if (eventRange.contains(curRange)) {
+                    timeranges.remove(i);
+                // case 2: event timerange is within current timerange
+                } else if (curRange.contains(eventRange)) {
+                    int numNewRanges = 0;
+                    if (eventRange.start() - curRange.start() >= request.getDuration()) {
+                        timeranges.add(i + (++numNewRanges), TimeRange.fromStartEnd(curRange.start(), eventRange.start(), false));                    }
+                    if (curRange.end() - eventRange.end() >= request.getDuration()) {
+                    timeranges.add(i + (++numNewRanges), TimeRange.fromStartEnd(eventRange.end(), curRange.end(), false));
+                    }
```

Comment:
Nit: make sure indentation and curly braces placement is consistent in this chunk here.